inlineplz/linter_runner.py

```diff
@@ -0,0 +1,389 @@
+# -*- coding: utf-8 -*-
+# pylint: disable=ungrouped-imports,broad-except,too-many-instance-attributes
+"""Linter configurations."""
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import asyncio
+import fnmatch
+import multiprocessing
+import os.path
+import shutil
+import sys
+import time
+import traceback
+
+from identify import identify
+import inlineplz.linters
+from inlineplz import message
+from inlineplz import registry
+from inlineplz.util import system
+
+
+class LinterRunner:
+    def __init__(
+        self,
+        install=False,
+        autorun=False,
+        ignore_paths=None,
+        config_dir=None,
+        enabled_linters=None,
+        disabled_linters=None,
+        trusted=False
+    ):
+        # TODO: Break this class down with composition or something
+        self.install = install
+        self.autorun = autorun
+        self.ignore_paths = ignore_paths or []
+        self.config_dir = config_dir or os.getcwd()
+        self.config_dir = os.path.abspath(self.config_dir)
+        self.fallback_config_dir = os.path.abspath(os.path.join(os.path.dirname(inlineplz.linters.__file__), "config"))
+        self.enabled_linters = enabled_linters or []
+        self.disabled_linters = disabled_linters or []
+        self.trusted = trusted
+
+        self.all_filenames = self.all_filenames_in_dir()
+
+        # Keep track of the parsed messages from all of the linters we're running
+        self.messages = message.Messages()
+
+        # track commands we've already run so that we don't re-run them
+        self.previous_install_commands = []
+
+        # Default event loop is SelectorEventLoop, which on Windows does not support subprocesses
+        if sys.platform == "win32":
+            self.event_loop = asyncio.ProactorEventLoop()
+            asyncio.set_event_loop(self.event_loop)
+        else:
+            self.event_loop = asyncio.get_event_loop()
+
+        # Limit the number of concurrent subprocesses we can run to something close to the number of CPUs we have
+        self.process_limiter = asyncio.Semaphore(value=multiprocessing.cpu_count(), loop=self.event_loop)
+
+    async def run_command(self, command, timeout=120):  # noqa: MC0001
+        print('Running command: {}'.format(" ".join(command)))
+        sys.stdout.flush()
+
+        popen_kwargs = {
+            "stdout": asyncio.subprocess.PIPE,
+            "stderr": asyncio.subprocess.STDOUT,
```

Comment:
why is stderr getting redirected to stdout instead of to PIPE?

Reply:
oh, is this so that everything ends up in stdout and we can just wait on stdout.readline?

Reply:
Yup, simplifies our flow since we just combine them together anyway, and now keeps them time-relative.

Reply:
nice